pyhap/characteristic.py

```diff
@@ -137,13 +165,10 @@ def set_value(self, value, should_notify=True, should_callback=True):
         :raise ValueError: When the value being assigned is not one of the valid values
             for this Characteristic.
         """
-        if self.properties.get('ValidValues') and \
-                value not in self.properties['ValidValues'].values():
-            raise ValueError
-        self.value = value
-        if self.setter_callback is not None and should_callback:
-            self.setter_callback(value)
-        if should_notify and self.broker is not None:
+        self.value = self._validate_value(value)
+        if self.setter_callback and should_callback:
+            self.setter_callback(self.value)
```

Comment:
I think the callback should just be a callback and only be called if the value is changed in HomeKit.

Reply:
@cdce8p could you point me to an example where `setter_callback` is used? It strikes me as odd that `setter_callback` is properly exposed or documented. Also I would be inclined to agree with you with regards to not call it.

Also `set_value` is an interesting name. In JS-HAP, there are two APIs `set_value` and `update_value`. `set_value` does not notify anything, just updates the objects internal state. Where as `update_value` will make sure that all the clients are notified of the change.

Reply:
To expand a little more on the above. There doesn't currently appear to be away for an `Accessory` to indicate a value update success or failure.

@ikalchev do you have any docs that explaining how client value updates are processed and forwarded onto `Accessory`s? I would be interested in extending it to allow the `Accessory` objects to indicate failure, and have that sent back to the requesting client.

Reply:
If I may: You have a bulb, which eventually the user taps on in iOS. This sends a request to HAP-python, which triggers the On characteristic's `set_value` method with `should_callback` as True. The `setter_callback` allows users of HAP-python to plug in and actually turn the bulb on, whether it is by interacting with GPIO or whatever.

About the docs: I haven't documented it, sorry about that. Will update soon.

In HAP-python, `update_value` is the JS `set_value`, whereas internal state can be modified with `self.value`.

Reply:
@ikalchev ok, if I understand this right.

iOS clients cause `set_value` to be called with `should_callback` as `True`, and `Accessory` objects should call `set_value` with `should_callback=False` and `should_notify=True`?

If that's the case wouldn't providing separate APIs be a little cleaner, rather than using flags?

Reply:
@thomaspurchas Client updates end up in `AccessoryDriver.set_characteristics`, which is where `char.set_value` is called. Currently, there is no (easy) way to indicate failure of this. I can propose two things:
1. `set_value` returns True or False, which is checked in `set_characteristics`
2. `set_value` throws something , which is catches in `set_characteristics`

In both cases the user should be able to plug into `set_value`. So, again, here are a few options:
- Use the current setter_callback - not perfect because the `self.value` is set before this is called. We can of course change the flow.
- Use some other mechanism

Personally, I think changing the flow in `set_value` as
```
...
success = self.callback(new_value)
if not success:
    return False
# continue as before
```
will be flexible enough, but some input will mean a lot. Thoughts?

Reply:
> Accessory objects should call set_value with should_callback=False and should_notify=True?

It really depends on your use case to be honest. I had discussion with people for this before, but we couldn't figure a better way for it. For example, if your accessory wants to turn on the light, it can just call `set_value(new_value)` to whatever is necessary (with `should_callback=True, should_notify=True`, both are the default) and then leave it to the `setter_callback` to actually turn the bulb on. On the other hand, if turning the lights CAN ALSO happen by some external action that you detect after it's done, then the above approach breaks down, because you have to have more complicated control logic.

The problem is how to handle both cases above well enough. So far I have found these flags acceptable in the cases I've seen, but then what have I seen? If you think the API is missing stuff or if you have a suggestion, I would be more than happy to discuss it and make changes.

Reply:
I will try to come up with a better way for this once more, but a use case/workflow to think about would be nice.

Reply:
The approach I would use is have `set_value` replaced with `update_value`, which should be called when an iOS client wants the value to change. Then the callback is responsible for trying to make that happen, and raises and exception if it can't.

For the `Accessory` object, I think that a `@property` setter could be used on the `value` attribute, that triggers the notification process.

I think it's very odd for people to be using this API to have the callback triggered. Presumably anything that wants to manipulate the `Characteristic` will do it via the `Accessory` object. In which case the `Accessory` object would be responsible for calling the callback, or only setting `value` once the callback has completed successfully.

That way there is a clear separation of responsibilities, `Characteristic`s only deal with what needs to be presented to iOS. Marshalling `value` sets to iOS clients as notifications, and marshalling requests from iOS to the `Accessory`. The `Accessory` is then responsible for making sure that the `value` stores in a `Characteristic` matches the real world device, regardless how that value was set.

This would also closely match how JS-HAP works as well, and prevents py-HAP from being tangled into peoples accessory code.

Happy to put together some cases/workflows, and also a PR to make that happen. Thought from yourself and @cdce8p welcomed!